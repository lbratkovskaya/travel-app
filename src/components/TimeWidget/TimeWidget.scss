@import '../widgets.scss';

.time-widget {
  @include widget;

  .time-widget-icon {
    color: $icon-color;
  }

  .time-widget-data {
    display: flex;
    flex-direction: column;
  }
}
