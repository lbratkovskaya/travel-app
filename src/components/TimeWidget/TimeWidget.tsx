import { AccessTimeOutlined } from '@material-ui/icons';
import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import rootConnector, { rootProps } from '../../store/rootConnector';
import { Country, URLParamTypes } from '../../types';
import './TimeWidget.scss';

const TimeWidget: React.FC<rootProps> = (props: rootProps) => {
  const { countryId } = useParams<URLParamTypes>();
  const [localTime, setLocalTime] = useState<Date>(new Date());
  const { lang, countries } = props;

  const currentCountry: Country | undefined = countries && countries
    .find((country: Country) => country.id === countryId);
  const capitalTimeZone: string | undefined = currentCountry?.capitalTimeZone;

  useEffect(() => {
    const timer1 = setInterval(() => {
      setLocalTime(new Date());
    }, 1000);

    return () => clearInterval(timer1);
  }, [currentCountry]);

  const timeZone = { timeZone: capitalTimeZone || 'UTC' } as Intl.DateTimeFormatOptions;

  const timeOptions = {
    hour: 'numeric',
    minute: 'numeric',
    second: 'numeric',
    ...timeZone,
  } as Intl.DateTimeFormatOptions;

  return (
    <div className="time-widget">
      <div className="time-widget-icon">
        <AccessTimeOutlined />
      </div>
      <div className="time-widget-data">
        <span>{`${localTime?.toLocaleDateString(lang, timeZone)}`}</span>
        <span>{`${localTime?.toLocaleTimeString(lang, timeOptions)}`}</span>
      </div>
    </div>
  );
};

export default rootConnector(TimeWidget);
