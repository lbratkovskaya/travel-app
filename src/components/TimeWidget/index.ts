import TimeWidget from './TimeWidget';

export default TimeWidget;
