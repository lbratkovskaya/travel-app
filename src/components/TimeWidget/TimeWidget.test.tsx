import React from 'react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { act } from 'react-dom/test-utils';
import { createMemoryHistory } from 'history';
import { mount, shallow, ShallowWrapper } from 'enzyme';
import TimeWidget from './TimeWidget';

const mockStore = configureStore([]);

const runAllPromises = () => new Promise(setImmediate);

describe('TimeWidget functional component testing', () => {
  let wrapper: ShallowWrapper<typeof TimeWidget>;

  const initialState = {
    countries: [
      {
        id: 'test',
        capitalTimeZone: 'Europe/Moscow',
      },
    ],
    lang: 'en',
  };

  const store = mockStore(initialState);

  beforeEach(() => {
    wrapper = shallow<typeof TimeWidget>(<Provider store={store}><TimeWidget /></Provider>);
  });

  it('TimeWidget must have a correct snapshot', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('TimeWidget must show location\'s date', async () => {
    const history = createMemoryHistory({
      initialEntries: ['/country/test'],
    });

    const component = mount(
      <Provider store={store}>
        <Router history={history}>
          <Route path="/country/:countryId">
            <TimeWidget />
          </Route>
        </Router>
      </Provider>,
    );

    await act(async () => {
      await runAllPromises();
      component.update();
      const nowDate = new Date();
      const expectedTest = nowDate.toLocaleDateString('en', { timeZone: 'Europe/Moscow' });
      expect(component.find('.time-widget-data').childAt(0).text()).toBe(expectedTest);
    });
  });
});
