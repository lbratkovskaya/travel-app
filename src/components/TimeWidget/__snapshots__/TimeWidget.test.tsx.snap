// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeWidget functional component testing TimeWidget must have a correct snapshot 1`] = `
"<ContextProvider value={{...}}>
  <Connect(TimeWidget) />
</ContextProvider>"
`;
